runtime/doc/develop.txt

```diff
@@ -166,77 +166,282 @@ stick to these rules, to keep the sources readable and maintainable.
-   version in sync with the main code base (it may be a while before your
-   changes will be included).  You do need to spend some time learning git,
-   it's not the most user friendly tool.
-2. Adjust the documentation.  Doing this first gives you an impression of how
-   your changes affect the user.
-3. Make the source code changes.
-4. Check ../doc/todo.txt if the change affects any listed item.
-5. Make a patch with "git diff".  You can also create a pull request on
-   github, but it's the diff that matters.
-6. Make a note about what changed, preferably mentioning the problem and the
-   solution.  Send an email to the |vim-dev| maillist with an explanation and
-   include the diff. Or create a pull request on github.
-
-
 C COMPILER				*style-compiler* *ANSI-C* *C89* *C99*
 
 The minimal C compiler version supported is C89, also known as ANSI C.
 Later standards, such as C99, are not widely supported, or at least not 100%
 supported.  Therefore we use only some of the C99 features and explicitly
 disallow some (this will gradually be adjusted over time).
 
-Please don't make changes everywhere to use the C99 features, it causes merge
-problems for existing patches.  Only use them for new and changed code.
+Features not to be used ~
 
-Comments ~
+These C99 features are not to be used, because not enough compilers support
+them:
+- Variable length arrays (even in C11 this is an optional feature).
+- C99 _Bool and _Complex types.
+- "inline" (it's hardly ever needed, let the optimizer do its work)
+- flexible array members: Not supported by HP-UX C compiler (John Marriott)
 
-Traditionally Vim uses /* comments */.  We intend to keep it that way
-for file and function headers and larger blocks of code, E.g.:
+COMMENTS					       *style-comments*
+
+Comments are good, but there is also a danger of over-commenting. NEVER try to
+explain HOW your code works in a comment. Write the code so that the working
+is obvious.
+
+Generally, you want your comments to tell WHAT your code does and WHY it does
+it, instead of HOW it is done.
+
+Try to avoid putting multiline comments inside a function body: if the
+function is so complex that you need to separately comment parts of it, you
+should probably rethink the structure of the function.
+
+Avoid using a comment to explain what a variable is, instead of naming the
+variable sensibly.
+
+For multiline comments use:
+>
+    /*
+     * comment
+     */
+<
+For inline comments use:
+>
+    int var; // comment
+<
+
+
+INDENTATION					      *style-indentation*
+
+We use 4 space to indent the code. If you are using vim, to edit the source,
+you don't need to do anything due to the |modeline|.
+
+For other editors an editorconfig is provided at the root of the repo.
+
+
+DECLARATIONS	                                     *style-declarations*
+
+Declare for loop variable in the guard (see also |style-loops|):
+Ok:
+>
+    for (int i = 0; i < len; ++i)
+<
+Wrong:
+>
+    int i;
+    for (i = 0; i < len; ++i)
```

Comment:
There are many places in the code where the loop variable is referenced outside of the for loop.  In those cases, obviously you need to declare the variable outside of the for loop.

Reply:
This should be changed especially if the variable is only used inside the loop. For other types of loops a while loop would be better.

Reply:
I believe this is mainly, because most of Vims source was written with C89 in mind, where this was not possible. It's a relative recent change to the style guide that we allowed to use this (I think when we in part switched to allow for C99)